### PR TITLE
ci: cancel superseded Full CI runs per PR

### PR DESCRIPTION
**Pull request by jasonqinzhou**
## Why and what changed

Full CI currently lets superseded PR revisions keep consuming runners. A second push to `pull-request/159`, for example, can leave both revisions' build/test matrices active.

Add workflow-level concurrency that replaces older queued and running validation for the same trusted PR copy. Automatic and manual runs on `pull-request/N` share a group. Manual source-branch runs replace only that branch's runs. Different PRs remain independent, and every main/release/tag run keeps a unique group.

## Review map

- Risk level: **Medium** — changes cancellation of active CI work.
- Start with: `.github/workflows/ci.yml` (group selection and cancellation); `tests/test_ci_workflow_contracts.py` (group identity and protected refs); `docs/ci.md` (current-SHA PR admission, manual-run limitations, and retry guidance).
- Public or serialized contract changed: No product API/schema changes. Full CI now cancels earlier runs in the same eligible ref group. Required check names, permissions, selection, and coverage are unchanged.
- Compatibility or rollback concern: Existing manual callers remain valid. Trusted-copy manual runs share the automatic cancellation group, but GitHub excludes manual job checks from required PR rulesets. Preserve eligible push-triggered validation; manual source-branch runs provide diagnostic evidence. Revert the concurrency block to restore independent runs. The policy applies only after branches incorporate this workflow; existing runs are not retroactively canceled. An old-revision rerun can replace a newer run, so replacement validation should target the current PR head.

## Evidence

Validated head: `72c3e703ae01c3b69f4a625afcdc972d916cf4db`, incorporating main `7ae483e942b0a2593d55e4e30957f94640580441`. Squash-merged into main as `c8ba25e8ff9d7c17dbb3d980e4c2311071e0c2bd` on 2026-09-19 at 03:19:09 UTC. The three-file feature diff is unchanged by the integration merge.

- Local tests: **316 passed** on this head with Python 3.12 and the bundled Node runtime. Command: `python -m pytest -c /dev/null tests/test_ci_workflow_contracts.py tests/test_ci_qualification.py tests/test_release_fpe.py -q -p no:cacheprovider --override-ini=filterwarnings=error`. The environment includes the CI-pinned `pip-licenses==5.5.5` dependency.
- Current-head checks: Ruff lint and format, copyright headers, packaged legal files, local documentation links, and `git diff --check` pass.
- Codex review: fresh complete-diff and surrounding-workflow review at `72c3e703ae01c3b69f4a625afcdc972d916cf4db` found no actionable issue. Both previous findings remain fixed and their threads are resolved.
- CodeRabbit: [current-head review](https://github.com/ai-dynamo/aisimulate/pull/294#issuecomment-5738873548) found no actionable issue. The bot disclosed its local history limitation; the full feature diff and main integration were independently verified locally.
- Fast CI: [PR run 35417501128](https://github.com/ai-dynamo/aisimulate/actions/runs/35417501128) passed on this head; codeowners and DCO also pass.
- Trusted merge-gate CI: [Fast CI 35417629984](https://github.com/ai-dynamo/aisimulate/actions/runs/35417629984) and [Full CI 35417630105](https://github.com/ai-dynamo/aisimulate/actions/runs/35417630105) both passed after exact-head admission. Full CI completed **54 successful jobs** with one expected PR release-staging skip. `pull-request/294` was verified to equal `72c3e703ae01c3b69f4a625afcdc972d916cf4db`. The earlier green runs are historical evidence only.
- Merge verification: applicable human approval was retained, both review threads are resolved, and required Fast CI, Full CI and codeowners checks passed. The resulting squash commit is on `main`, has parent `7ae483e942b0a2593d55e4e30957f94640580441`, and exactly matches the validated tree `8c80e62912b7a3c5ceb30c088c84b8e19f66d12c`; `git diff` between validated head and merge commit is empty.
- Negative or boundary cases: same-PR push/manual identity; different PR isolation; same versus different manual source branches; unique main, release, and tag groups for both push and manual events.
- Expected-value derivation (for exact numerical assertions): N/A — no prediction/numerical assertions changed.
- Before/after output, trace, benchmark, or golden diff: Before, both PR revisions could remain active. After, native Actions concurrency gives the same PR ref one shared cancellation group. Tests evaluate the checked-in expression; scheduler cancellation itself has not been exercised with a deliberate live duplicate run.

## Modeling or data provenance

N/A — no modeling, performance data, or prediction changes.

## Tracking

- Closes: N/A.
- Related PRs or issues: N/A; follow-up to the observed AISim AMD64 CI queue and concurrent superseded PR runs.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -25,6 +30,18 @@ permissions:
...
 
+concurrency:
+  # PR copies share a group across push/manual events. Source-branch manual
+  # runs can join that group with pr_number; otherwise they replace their own
+  # branch's runs. Unique lifecycle/tag groups preserve every staging run.
+  group: >-
+    full-ci-${{ startsWith(github.ref, 'refs/heads/pull-request/') && github.ref ||
+      (github.event_name == 'workflow_dispatch' && startsWith(github.ref, 'refs/heads/') &&
+       github.ref != 'refs/heads/main' && !startsWith(github.ref, 'refs/heads/release/')) &&
+      (inputs.pr_number && format('refs/heads/pull-request/{0}', inputs.pr_number) || github.ref) ||
```
Comment (jasonqinzhou) [inline .github/workflows/ci.yml]: [P1] Validate PR identity before choosing a cancellation target. Workflow-level concurrency takes effect before Verify manual PR identity runs. A manual dispatch on the source branch for PR #159 with pr_number=160 therefore cancels valid PR #160 CI, then fails its identity check. Reproduced by evaluating the checked-in group expression for both refs: they both produce full-ci-refs/heads/pull-request/160. Remove the unvalidated alias input and use the canonical trusted-copy ref for shared manual/automatic cancellation; retain branch-scoped groups for source-branch callers.

Comment (jasonqinzhou) [reply .github/workflows/ci.yml]: Fixed in 5c23c24450b4bfbc59fa528958bc9b71cb2617b1. Removed the PR-number alias input and its too-late validation. Cancellation identity now comes only from the workflow ref; manual and automatic PR runs share the trusted pull-request/N ref. The guide documents exact-head verification before dispatch. All 316 CI contract tests and actionlint pass; fresh review of the published diff found no remaining actionable issue.

**Review by simone-chen — APPROVED** [review]

**Comment by jasonqinzhou** [comment]
@coderabbitai full review

Please review current head `56aeb08a48d4c0276c3724acffb8fa8382f3a076` after synchronizing with main. The net change remains the same three files. Local CI contract tests (316), actionlint, Ruff, and source/documentation checks pass; the fresh Codex review found no remaining actionable issue. Standalone Fast CI has passed on this commit.

File: docs/ci.md [hunk 2]
```diff
@@ -567,6 +568,32 @@ gh workflow run ci.yml --repo ai-dynamo/aisimulate \
   --ref "${ci_branch}" -f expected_sha="${ci_sha}"
 ```
 
+Full CI cancels older queued and running validation for the same trusted
+`pull-request/N` copy, including manual dispatches on that copy. To replace an
+automatic PR run manually, first verify that its trusted copy matches the PR's
+current head, then dispatch on the copy:
```
Comment (jasonqinzhou) [inline docs/ci.md]: [P2] Preserve a PR-eligible replacement check. GitHub does not evaluate workflow_dispatch job checks for required PR rulesets, even at the PR head (confirmed by manual Full CI run 35410750694 succeeding while this PR still lacks Full CI Success). This example would cancel an eligible trusted-copy push run and replace it with a run that cannot satisfy the gate. Replace the manual-replacement recommendation with the current-SHA trusted-copy admission path, and clearly label manual runs as diagnostic evidence. Source: https://docs.github.com/en/pull-requests/how-tos/merge-and-close-pull-requests/troubleshooting-required-status-checks#checks-from-some-workflow-jobs-are-not-evaluated

Comment (jasonqinzhou) [reply docs/ci.md]: Fixed in ddb63d5dd2ecb25a2fc37c7c72255b3a0a93e134. The guide now uses current-SHA copy-pr-bot admission for eligible push-triggered PR checks, identifies manual runs as diagnostic evidence, warns that trusted-copy manual dispatch can cancel an eligible push result, and directs transient failures to the existing push run. The concurrency implementation is unchanged.

**Review by jasonqinzhou — COMMENTED** [review]
Fresh Codex review of 72c3e703ae01c3b69f4a625afcdc972d916cf4db after integrating main 7ae483e942b0a2593d55e4e30957f94640580441: no remaining actionable finding.

Reviewed the full three-file feature diff and surrounding workflow dispatch, trusted-copy SHA verification, scope selection, Fast CI prerequisite, and release staging paths. The feature diff is byte-identical to the prior revision. Cancellation identity uses only the workflow ref; trusted copies group across push/manual events, source branches are isolated by ref, and lifecycle/tag runs retain unique run IDs. Both prior inline findings are fixed and resolved.

Validation on this head: 316 CI/workflow/release tests passed with Python 3.12, working Node, and CI-pinned pip-licenses; Ruff lint/format, copyright headers, packaged legal files, documentation links, and diff hygiene passed.

The CodeRabbit modeling-data warning does not identify a product-model change: this PR changes GitHub scheduling policy only. The tests provide explicit cancellation/isolation expectations for eight boundary cases, while the description states that deliberate live duplicate-run cancellation is not exercised. Native scheduler semantics match GitHub's documented concurrency contract. No numerical oracle or modeling fixture is applicable.

Current-head CodeRabbit and required CI remain to be refreshed. Human approval is separate and will be rechecked before merge.

**Comment by jasonqinzhou** [comment]
@coderabbitai review current head 72c3e703ae01c3b69f4a625afcdc972d916cf4db. Main is now integrated; the complete three-file feature diff is unchanged. Please refresh the current-head review before trusted Full CI admission.

**Comment by jasonqinzhou** [comment]
Verified squash merge into main: `c8ba25e8ff9d7c17dbb3d980e4c2311071e0c2bd` (2026-09-19 03:19:09 UTC).

- Reviewed/tested head: `72c3e703ae01c3b69f4a625afcdc972d916cf4db`.
- Exact-head [trusted Fast CI](https://github.com/ai-dynamo/aisimulate/actions/runs/35417629984) and [trusted Full CI](https://github.com/ai-dynamo/aisimulate/actions/runs/35417630105) passed. Full CI: 54 successful jobs and one expected release-staging skip for the PR branch. Required codeowners and PR Fast CI also passed.
- Applicable human approval remains recorded; all review threads are resolved. Current-head Codex and CodeRabbit reviews found no actionable issue.
- Fetched main and verified the merge commit is on it. Its parent is `7ae483e942b0a2593d55e4e30957f94640580441`; its tree exactly equals the validated head's `8c80e62912b7a3c5ceb30c088c84b8e19f66d12c`, with an empty head-to-merge diff.

PR description and validation notes now reflect the completed merge.
